modules/auxiliary/gather/enum_dns.rb

```diff
@@ -1,483 +1,420 @@
-              :sname => 'dns',
-              :port => 53 ,
-              :type => 'dns.enum',
-              :update => :unique_data,
-              :data => "#{rr.address.to_s},#{line.chomp}.#{target},AAAA")
-            next unless rr.class == Net::DNS::RR::CNAME
-          end
-        end
-      end
+  def get_ns(domain)
+    resp = dns_query(domain, 'NS')
+    return if resp.blank? || resp.answer.blank?
 
+    records = []
+    resp.answer.each do |r|
+      next unless r.class == Net::DNS::RR::NS
+      records << "#{r.nsdname}"
+      report_host(host: r.nsdname, name: domain, info: 'NS')
+      report_service(
+        host: r.nsdname,
+        name: 'dns',
+        port: 53,
+        proto: 'udp',
+        info: 'nameserver')
+      vprint_good("#{domain}: NS: #{r.nsdname}")
     end
+    return if records.none?
+    path = store_loot(
+      'ENUM_NS',
+      'text/plain',
+      domain,
+      records,
+      'ENUM_NS')
+    print_good('Saved file to: ' + path)
+    records
   end
 
+  def get_mx(domain)
+    resp = dns_query(domain, 'MX')
+    return if resp.blank? || resp.answer.blank?
 
-
-  def reverselkp(iprange,nssrv)
-    print_status("Running reverse lookup against IP range #{iprange}")
-    if not nssrv.nil?
-      @res.nameserver = (nssrv)
-      @nsinuse = nssrv
-    end
-    ar = Rex::Socket::RangeWalker.new(iprange)
-    tl = []
-    while (true)
-      # Spawn threads for each host
-      while (tl.length < @threadnum)
-        ip = ar.next_ip
-        break if not ip
-        tl << framework.threads.spawn("Module(#{self.refname})-#{ip}", false, ip.dup) do |tip|
-          begin
-            query = @res.query(tip)
-            raise ::Rex::ConnectionError
-            query.each_ptr do |addresstp|
-              print_status("Hostname: #{addresstp} IP address: #{tip.to_s}")
-              report_note(:host => @nsinuse.to_s,
-                :proto => 'udp',
-                :sname => 'dns',
-                :port => 53 ,
-                :type => 'dns.enum',
-                :update => :unique_data,
-                :data => "#{addresstp},#{tip},A")
-            end
-          rescue ::Interrupt
-            raise $!
-          rescue ::Rex::ConnectionError
-          rescue ::Exception => e
-            print_error("Error: #{tip}: #{e.message}")
-            elog("Error running against host #{tip}: #{e.message}\n#{e.backtrace.join("\n")}")
-          end
-        end
-      end
-      # Exit once we run out of hosts
-      if(tl.length == 0)
-        break
-      end
-      tl.first.join
-      tl.delete_if { |t| not t.alive? }
+    records = []
+    resp.answer.each do |r|
+      next unless r.class == Net::DNS::RR::MX
+      records << "#{r.exchange}"
+      report_host(host: r.exchange, name: domain, info: 'MX')
+      report_service(host: r.exchange, name: 'smtp', port: 25, proto: 'tcp')
+      vprint_good("#{domain}: MX: #{r.exchange}")
     end
+    return if records.none?
+    path = store_loot(
```

Comment:
When run, `loot` shows this:

```
host             service  type      name      content     info  path
----             -------  ----      ----      -------     ----  ----
205.251.203.234           ENUM_NS   ENUM_NS   text/plain        /home/jhart/.msf4/loot/20151228093339_default_205.251.203.234_ENUM_NS_258338.txt
205.251.203.236           ENUM_TXT  ENUM_TXT  text/plain        /home/jhart/.msf4/loot/20151228093411_default_205.251.203.236_ENUM_TXT_932411.txt
205.251.203.237           ENUM_NS   ENUM_NS   text/plain        /home/jhart/.msf4/loot/20151228093410_default_205.251.203.237_ENUM_NS_589122.txt
205.251.203.238           ENUM_SOA  ENUM_SOA  text/plain        /home/jhart/.msf4/loot/20151228093411_default_205.251.203.238_ENUM_SOA_632335.txt
205.251.203.238           ENUM_MX   ENUM_MX   text/plain        /home/jhart/.msf4/loot/20151228093340_default_205.251.203.238_ENUM_MX_250988.txt
205.251.203.240           ENUM_NS   ENUM_NS   text/plain        /home/jhart/.msf4/loot/20151228093415_default_205.251.203.240_ENUM_NS_261853.txt
205.251.203.247           ENUM_MX   ENUM_MX   text/plain        /home/jhart/.msf4/loot/20151228093411_default_205.251.203.247_ENUM_MX_765527.txt
205.251.203.247           ENUM_TXT  ENUM_TXT  text/plain        /home/jhart/.msf4/loot/20151228093340_default_205.251.203.247_ENUM_TXT_759378.txt
205.251.203.247           ENUM_SOA  ENUM_SOA  text/plain        /home/jhart/.msf4/loot/20151228093340_default_205.251.203.247_ENUM_SOA_332543.txt
205.251.203.249           ENUM_NS   ENUM_NS   text/plain        /home/jhart/.msf4/loot/20151228093345_default_205.251.203.249_ENUM_NS_858128.txt
```

Several problems:
 * The `ENUM_*` stuff is duplicated
 * There is no indication what domain this is for
 * The content is a CSV, not text (change the extension or type?)